torch/_subclasses/functional_tensor.py

```diff
@@ -547,16 +592,51 @@ def unwrap(x):
                         *args_unwrapped,
                         **kwargs_unwrapped,
                     )
-                    # We don't allow any mutation on result of dropout or _to_copy
-                    if self.export:
-                        if func in (
-                            torch.ops.aten.dropout.default,
-                            torch.ops.aten._to_copy.default,
-                        ):
-                            torch._freeze_functional_tensor(outs_unwrapped)  # type: ignore[attr-defined]
+
                     outs_wrapped = pytree.tree_map_only(
                         torch.Tensor, wrap, outs_unwrapped
                     )
+
+                    if self.export:
+                        # check all args. If any of them is an alias to the result of aten.to and is mutated,
+                        # we mark it in _mutated_partial_frozen_storage.
+                        for arg_id, arg in enumerate(args):
+                            if isinstance(
+                                arg, FunctionalTensor
+                            ) and torch._is_functional_tensor(arg.elem):
+                                arg_storage = StorageWeakRef(arg.untyped_storage())
+                                arg_mutation_counter = torch._mutation_counter(arg.elem)  # type: ignore[attr-defined]
+                                if arg_mutation_counter > arg_mutation_counters[arg_id]:  # type: ignore[index]
+                                    # store tensor aliases that's mutated after it's partial frozen
+                                    arg_group_id = self._partial_frozen_storage.get(
+                                        arg_storage, None
+                                    )
+                                    if arg_group_id is not None:
+                                        self._mutated_partial_frozen_storage.add(
+                                            arg_group_id
+                                        )
+
+                        # TODO: allow it in draft export mode and attach to FailureReport
+                        if func == torch.ops.aten._to_copy.default:
```

Comment:
Since we are only ever doing this logic for `aten._to_copy`... do we get the behavior that we want in the case that `.to()` becomes a no-op?

I think it would be good to have a test that we error for e.g. this:
```
def f(x):
    y = x + 1
    # fp32.to(fp32) becomes a no-op
    z = y.to(dtype=torch.float32)
    z.mul_(2)
    # this should error
    return y
    
x = torch.randn(4, dtype=torch.float32)
gm = export(f, (x,))
```

Reply:
I think this test case is a no-op? I can include a test for your case as well. 


```
        class Module(torch.nn.Module):
            def forward(self, x):
                y = x.to("cpu")
                y.add_(1)
                l = x + 1
                return l

        inps = (torch.tensor(1, device="cpu"),)
```


My understanding is even if `.to()` is a no-op, it still get's traced, and it will be re-dispatched as `aten._to_copy`. here: https://github.com/pytorch/pytorch/blob/80761c62f47df55d6f7c231d23c5fcacc3c0a7f0/torch/_subclasses/functional_tensor.py#L401--L413

Reply:
If `x.to(...)` is a no-op, then it will just return `x` directly without dispatching to `_to_copy` (here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorConversions.cpp#L429)

Reply:
why is the test above a no-op?

Reply:
> why is the test above a no-op?

I thought it's a no-op because the input is already in "cpu", so `.to("cpu")` is a no-op. Maybe I'm wrong?

Reply:
oh yep, the `.to()` call becomes a no-op when running that code. But I would still expected us to error when attempting to export that code: depending on whether or not you pass a cpu or a cuda tensor to that function, you will get different results.

Reply:
> oh yep, the `.to()` call becomes a no-op when running that code. But I would still expected us to error when attempting to export that code: depending on whether or not you pass a cpu or a cuda tensor to that function, you will get different results.

sounds good. Yes, we do error when attempting to export that code. 

> If x.to(...) is a no-op, then it will just return x directly without dispatching to _to_copy (here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorConversions.cpp#L429)

We still dispatch to _to_copy because we have a custom `to()` implemented for FunctionalTensor:

https://github.com/pytorch/pytorch/blob/80761c62f47df55d6f7c231d23c5fcacc3c0a7f0/torch/_subclasses/functional_tensor.py#L291

Reply:
oh fwiw @tugsbayasgalan - this won't catch any `tensor.to()` calls that happen from C++

Reply:
I guess in the long term this is fine because the recommended way to go inference IR is via training IR where we would have materialized aten.to in the first place.

Reply:
Just to give a counterexample: you can have a CIA op that shows up directly in training IR, that has a C++ implementation that calls `TensorImpl::to()`